File: MiniDB_Projects/Team_stormborns/src/BufferPool.cpp
#include "BufferPool.h"

#include <stdexcept>

BufferPool::BufferPool(DiskManager* diskManager, int capacity)
    : diskManager_(diskManager), capacity_(capacity)
{
    if (diskManager == nullptr) {
        throw std::invalid_argument("DiskManager must not be null");
    }
    if (capacity <= 0) {
        throw std::invalid_argument("Capacity must be > 0");
    }
}

BufferPool::BufferPool(DiskManager* diskManager)
    : BufferPool(diskManager, DEFAULT_CAPACITY)
{
}

BufferPool::~BufferPool() {
    flushAllPages();

    // Explicitly delete all heap-allocated Page objects
    for (auto& pair : pageTable_) {
        delete pair.second;
    }
    pageTable_.clear();
    lruList_.clear();
    lruMap_.clear();
    pinCounts_.clear();
}

Page* BufferPool::fetchPage(int pageId) {
    // Check if page is already in the cache
    auto it = pageTable_.find(pageId);

    if (it != pageTable_.end()) {
        // Cache hit — promote in LRU
        promoteLRU(pageId);
        pinCounts_[pageId]++;
        return it->second;
    }

    // Cache miss — may need to evict
    if (static_cast<int>(pageTable_.size()) >= capacity_) {
        evictPage();
    }

    // Read page from disk into a new heap-allocated Page
    uint8_t buf[Page::PAGE_SIZE];
    diskManager_->readPage(pageId, buf);

    Page* page = new Page(pageId, buf);
    pageTable_[pageId] = page;

    // Add to front of LRU list (most recently used)
    lruList_.push_front(pageId);
    lruMap_[pageId] = lruList_.begin();

    // Pin the page
    pinCounts_[pageId] = 1;

    return page;
}

void BufferPool::unpinPage(int pageId, bool isDirty) {
    auto it = pageTable_.find(pageId);
    if (it == pageTable_.end()) {
        return;
    }

    if (isDirty) {
        it->second->markDirty();
    }

    int currentPins = pinCounts_[pageId];
    if (currentPins > 0) {
        pinCounts_[pageId] = currentPins - 1;
    }
}

void BufferPool::evictPage() {
    // Scan from back (least recently used) to find an unpinned page
    for (auto rit = lruList_.rbegin(); rit != lruList_.rend(); ++rit) {
        int candidateId = *rit;

        if (pinCounts_[candidateId] == 0) {
            // Found an evictable page
            Page* evictPage = pageTable_[candidateId];

            // Write to disk if dirty
            if (evictPage->isDirty()) {
                diskManager_->writePage(evictPage->getPageId(),
                                        evictPage->getData());
            }

            // Remove from all data structures
            pageTable_.erase(candidateId);
            pinCounts_.erase(candidateId);

            // Remove from LRU list using the stored iterator
            auto listIt = lruMap_[candidateId];
            lruList_.erase(listIt);
            lruMap_.erase(candidateId);

            // Free the Page memory
            delete evictPage;
            return;
        }
    }

    throw std::runtime_error(
        "Buffer pool is full and all pages are pinned!");
}

void BufferPool::promoteLRU(int pageId) {
    auto mapIt = lruMap_.find(pageId);
    if (mapIt == lruMap_.end()) return;

    // Move to front of LRU list
    lruList_.erase(mapIt->second);
    lruList_.push_front(pageId);
    lruMap_[pageId] = lruList_.begin();
}

void BufferPool::flushAllPages() {
    for (auto& pair : pageTable_) {
        Page* page = pair.second;
        if (page->isDirty()) {
            diskManager_->writePage(page->getPageId(), page->getData());
            page->clearDirty();
        }
    }
}

File: MiniDB_Projects/Team_23bcs10071/src/minidb/buffer_pool.py
from collections import OrderedDict

class BufferPool:
    def __init__(self, capacity=16):
        self.capacity = capacity
        self.pages = OrderedDict()  # page_no -> (dirty, data)

    def get(self, page_no, load_fn, write_fn=None):
        if page_no in self.pages:
            # move to end = most recently used
            self.pages.move_to_end(page_no)
            return self.pages[page_no][1]
        page = load_fn(page_no)
        if len(self.pages) >= self.capacity:
            # evict least recently used
            evict_page_no, (dirty, pdata) = self.pages.popitem(last=False)
            if dirty and write_fn:
                write_fn(evict_page_no, pdata)
        self.pages[page_no] = (False, page)
        return page

    def mark_dirty(self, page_no):
        if page_no in self.pages:
            dirty, page = self.pages[page_no]
            self.pages[page_no] = (True, page)

    def put(self, page_no, page, dirty=False):
        self.pages[page_no] = (dirty, page)
        self.pages.move_to_end(page_no)

    def flush_all(self, write_fn):
        for page_no, (dirty, page) in list(self.pages.items()):
            if dirty:
                write_fn(page_no, page)
                self.pages[page_no] = (False, page)
